### ISSUE TITLE
Permanent URL support

### ISSUE THREAD

**Issue by MrCoder** [chinchang/web-maker#295]
It would nice if I can click a perm-link to open a page. For example - https://webmakerapp.com/item/asdfasdfasdfasdfasdf.

**Comment by chinchang** [comment]
I get the point, but curious if in your case you need the URL for yourself or to share with someone else?

**Comment by MrCoder** [comment]
It's for sharing. In my mind is something like what codesandbox.io or jsfiddle has. One may share it on stackoverflow for example.

**Comment by neomodern** [comment]
I'd like the URL available just for personal use paired with the local-only saves.

**Comment by vanloc0301** [comment]
do you have the plan for this feature?

**Comment by MrCoder** [comment]
Re the last point, firebase libraries are cached the first time they are loaded and should be fast enough from then on. It is important that same process is followed for App and Embedded from my point of view.

**Comment by MrCoder** [comment]
My previous response is not very clear. Let me rephrase it. I think embedding and permanent link are two related functionalities. For example, on WordPress, there are sharable links and no embedding. From my point of view, point two and three are not valid. Just my two cents.

**Comment by david-shortman** [comment]
I think links could share data like the TS playground does, by encoding it in base64 or something in the url: https://www.typescriptlang.org/play?#code/PTAEBEEMDcEsBNQHdIGdQAsCmAnLQ

I might play around with a PR for this 👀

**Comment by chinchang** [comment]
@david-shortman That is definitely not base64 :) I think it is a usual ID generated to associate to some code stored in the backend. Also, storing code in the URL will have limit on the number of characters we can have in the URL.

**Comment by david-shortman** [comment]
@chinchang Whoops my bad, that isn't base64.

However, it is _some_ type of encoded data, as the URL increases in size the more code I put in:

https://www.typescriptlang.org/play?#code/KYDwDg9gTgLgBAYwgOwM7wEoFEDiBJAZQBUtsARAfQEEBhIvAeQDkKiBNABSwIC44BvOAG0AhghgBLFEQCeYYH3RQJyAOYBdPsgCuAWwBGwKHAC+cALwCTAbgBQt0JFiIU6OCKhQRM1FSjAmCBgqABsQiAB3YAATAFlUVQtbODgAcg8vH3d-OGQg9zDImLgVd3EpZER-ERhoVFS7GDlgOD9M339A4MKo6Is4JvkIADNs9r8AoNDw3vjVOwdwaHgkNHhB4A4oCHlYGTxULumiuISklNSNuDBt3aaS1Fz8kR7i0rFJFCrgGrqG2yusnkWx2RiaByOrz6lg2IwGzRBd32hymULmC0cyxcazgwF0YCaDH0ACtgOIOpNujMYnNzmk8QSZHAICSyTBHh4Wnl4C9qX13uUvghqrUoPVGs04Fh8YTWeSJpC+f1YaMGbLSfLOqi+ej7JjnFcAGLaZCC5AQ7UnAA8yTgRAANLasJ5oLFgKhUCJVC1QDBgMhoo8lCpVLYAHz9Ii4kB+gOPY2mz6VAD8UpdUDdHq9LT4RDsAHoAFSF5KF1pwYYms0DAAWNTg-hg2igaHclRZGvgpRgNZaqDr8mZox7LQABlQzaOSsg-VBhmJgAA6FI0FDDCSqZvFbSoENwUfCn5+idJ0eL0v5xZOdaSmgi6A2lIcaOxwNtmRCdT9ETID-qR0pBgL7+m+HZsv0YHiOG-QJmaADqEg9hwHgiLowCzqgQLAFaHD2nAGBhhiSwGpKionDQvYIAA1laUa+iBjyQTAEaWHRMYMe+n62qmbTeBSZG9LaubAXGAjwvIfA-kyJjcXaCK3GCyICTEQlwFRwAyHCbGvo8yDAAAbkYsnSoyRKdvxlqCSkfAmlReQRMgCwpEWJZwGWBDAGOd5HtAZ4XleWJXCeFTeb8UCPnaIlvsGaj9DFqgAXANBRY8oWiv0aXQNBljJQAZHJ8jRMFKC0YRerEV2M5GPOCAtMVyCZVAiKoKVAi2hQIioHwqQ3Ds4rtWAuZ2DJAUkYOsFJghSEoWhGFYThKVwLZ9nIJ+eFAZY+kQBI0QsXAAAUtqLsdHiqF1cAcLYACUFgRhgRHXtOs41XV1b8LaGyKDAyhqMN9j5vmcDRGSIScgMEBwLoIjqSU8C9UiAC0-jIKhe6esMwCjfAwMIKDOQqM9C4FTE9W0Yt8URvRon1W1KTVNEKAhEyn12n9tiVomFRA8A656fN2kcRTh0pCzDq2oeYV8I1118PVjWte9dPoc2lRmWyi7A7zmwKXs+0S6KeGXM0qR4YrKRwPpLzaAo4nAIlKQRMoMAiPoIQ2-OISoHbtomFdbP6vAHPVvrwCkwLolCx9zS5jLrRmvLDoHTd5gRlhRVmqVZUBxWVZJt8NShxn4fRd9IZ4c+VOgXKzHC7buaJas66qLL8f3k1twtRwEb5cp0QUWSNFd7Hctt7ReH7fD53Pj3lkxP31E4WGycRtPxPp0mmd2NnQd5yHpO2sXQal2oiUV+xolSWttpAZXjHV+Gteiw3bdS6PuH4d3cC9-PNEEbHE0VCmjWZCXhZpGEws0HC604D5TTmHMMn84FFyzhVHOnMhQikLhvQ+cB4p4WSrfJKbcwyP2jnaBua4NzJj4IIDq50eod1SKYOAAAfJKw9W4+XCkQCMZtsD4GIKQLAlBaD0GYKwTg3AhAbC-JYfa-DCAkHINQOgjAWDsC4AQaRzQvwsLYQABhugAajgAARiciUUY+0VTYibhYcwlhUg7wqKkG6ZslZNhbNzLWWFrHNHHsdRcp1zqX3UMvA67jzaBMbhufagTglXXtubDYSTfZ+x9uLVwPJHiWBiYkVMeTFx0LgN1NUMh-gpFQBERCCAawHU6m420KQECdRaKkMpqQeBNPNo2FW3iVDAF8SkpOt0Im21MFddJ5tECtLSJPTp3SPF9M1gMoZ-iDqTz4ExcJ+1InTMCZPJJ0yUmLImVM82msRDaBCDALp0yRY1m2BEXIwBnnOm2FAfaqQACqelwBsmKIU1xdgUgyRGqsNwDN3QzjgtAKi-Q95mi+R8Fx48RBfR+glOA+h+gey9jswQIg8I4rSXYKFqAYVwq+agCAs0awhmBbYcllKoBUWpbS9C9K1AmwGFAa26T7AQpWP0fa6LcHHyxTiyweLgAEvcMSiZdghVwAiHCnJ+djxIvSGaHlCB0mqtZagdldKGX6rVcazlDK8LfX5dYIAA

**Comment by david-shortman** [comment]
[A random StackOverflow answer](https://stackoverflow.com/questions/417142/what-is-the-maximum-length-of-a-url-in-different-browsers) says to keep URLs under 2000 characters, and my above example breaks that by a few hundred (but still works on my up-to-date Safari).

Probably for any sufficiently large code, a URL won't cut it, but for sharing snippets that are only a few hundred lines of code, it may work.

**Comment by david-shortman** [comment]
Probably more sharing use cases could be covered by https://github.com/chinchang/web-maker/issues/292, so I'll look at that, too

**Comment by chinchang** [comment]
You are right! We can at least have this for small snippets! Looking forward to your PR. This reminds me, I did a similar thing in another project where I compressed the content first before putting in the URL...that can gets us more code. You can refer that here -> https://github.com/chinchang/picssel-art/blob/master/picssel.js#L354

**Comment by hatemhosny** [comment]
First, I'd like to thank you for such a great and inspiring project.

I have recently implemented a similar feature in my project [LocalPen](https://github.com/hatemhosny/localpen). It was indeed similar to typescript playground sharing feature.

You can try it on the development build here (settings menu > share): 
https://dev.localpen.io/


An object representing the current state of the code is compressed using LZString and added to the URL as hash preceded by the string `code/`. The URL is copied to clipboard to be ready for sharing.
When the URL is loaded, if a hash containing the string `code/` is found, the following hash is decompressed and loaded as the state object.

share:
https://github.com/hatemhosny/localpen/blob/32c15e4c8fdfef8df306bddc529e52540bdf614a/src/localpen/app.ts#L548

load:
https://github.com/hatemhosny/localpen/blob/32c15e4c8fdfef8df306bddc529e52540bdf614a/src/localpen/import/code.ts

compress/decompress
https://github.com/hatemhosny/localpen/blob/32c15e4c8fdfef8df306bddc529e52540bdf614a/src/localpen/utils.ts#L93
